src/exe/Main.hs

```diff
@@ -225,17 +248,41 @@ processInputFile
   :: Members '[Report, Embed IO] r => Options -> FilePath -> Sem r ()
 processInputFile opts inputFile = do
   input <- embed $ readFile inputFile
-  let state       = initPMState opts
-      inputModule = HSE.fromParseResult (HSE.parseModule input)
-      intermediateModule =
-        evalPM (processModule (FromHSE.transformModule inputModule)) state
-      outputModule = ToHSE.transformModule inputModule intermediateModule
+  let state             = initPMState opts
+      (output, modName) = processInput input state (optFrontend opts)
+--      inputModule = HSE.fromParseResult (HSE.parseModule input
+--      intermediateModule =
+--        evalPM (processModule (FromHSE.transformModule inputModule)) state
+--      outputModule = ToHSE.transformModule inputModule intermediateModule
   case optOutputDir opts of
     Just outputDir -> do
-      let outputFile = outputDir </> makeOutputFileName inputFile inputModule
+      let outputFile = outputDir </> makeOutputFileName inputFile modName
       embed $ createDirectoryIfMissing True (takeDirectory outputFile)
-      embed $ writeFile outputFile (prettyPrintModule outputModule)
-    Nothing -> embed $ putStrLn (prettyPrintModule outputModule)
+      embed $ writeFile outputFile output -- (prettyPrintModule outputModule)
+    Nothing -> embed $ putStrLn output --(prettyPrintModule outputModule)
+
+-- | Processes a given module as a string with the corresponding frontend.
+--   Returns the output as a string and the module name.
+processInput
+  :: String -> PMState FromHSE.HSE -> Frontend -> (String, Maybe String)
+processInput input state frontend = case frontend of
+  HSE ->
+    let inputModule = HSE.fromParseResult (HSE.parseModule input)
+        intermediateModule =
+            evalPM (processModule (FromHSE.transformModule inputModule)) state
+        outputModule = ToHSE.transformModule inputModule intermediateModule
+    in  (prettyPrintModule outputModule, moduleName inputModule)
+  GHClib -> error "Not yet implemented"
+ where
+  -- | Gets the module name of the given module.
+  --
+  --   Returns @Nothing@ if there is no module header or the module type is
+  --   not supported.
+  moduleName :: HSE.Module l -> Maybe String
```

Comment:
This function name should include HSE, for example `moduleNameHSE`.

Reply:
This function can probably also be implemented on `S.Module` instead.

Reply:
It can't be implemented in the current version of `S.Module` because the constructor has no name parameters. I can add one if it is OK for you.

Reply:
> the constructor has no name parameters. I can add one if it is OK for you.

Sounds good to me.

Reply:
I'm okay with that as well.